File: Intelligent-Audio-TEST/api_gateway/schemas/task.py
from __future__ import annotations

from typing import Any, Dict, List, Optional
from pydantic import Field, field_validator, model_validator

from api_gateway.schemas.base import APIModel
from api_gateway.schemas.common import PaginatedData


class TaskReportItem(APIModel):
    id: int = Field(...)
    name: str = Field(...)
    status: str = Field(...)
    type: str = Field(...)
    created_at: Optional[str] = Field(None)


class TaskReportsData(APIModel):
    count: int = Field(...)
    reports: List[TaskReportItem] = Field(default_factory=list)


class TaskDeviceBrief(APIModel):
    id: int = Field(...)
    name: str = Field(...)
    status: Optional[str] = Field(None)
    model: Optional[str] = Field(None)


class TaskApiBrief(APIModel):
    id: int = Field(...)
    name: str = Field(...)
    status: Optional[str] = Field(None)


class TaskListItem(APIModel):
    id: int = Field(...)
    name: str = Field(...)
    description: Optional[str] = Field(None)
    status: str = Field(...)
    type: str = Field(...)
    config: Dict[str, Any] = Field(default_factory=dict)
    algorithm_type: Optional[str] = Field(None)
    algorithm_params: Optional[Dict[str, Any]] = Field(None)
    started_at: Optional[str] = Field(None)
    completed_at: Optional[str] = Field(None)
    total_cases: Optional[int] = Field(None)
    case_count: Optional[int] = Field(None)
    device_count: Optional[int] = Field(None)
    completed_cases: Optional[int] = Field(None)
    failed_cases: Optional[int] = Field(None)
    tags: List[str] = Field(default_factory=list)
    created_at: Optional[str] = Field(None)
    updated_at: Optional[str] = Field(None)
    reports: Optional[TaskReportsData] = Field(None)
    devices: List[TaskDeviceBrief] = Field(default_factory=list)
    apis: List[TaskApiBrief] = Field(default_factory=list)


class TaskListData(PaginatedData[TaskListItem]):
    pass


class TaskCaseBrief(APIModel):
    case_id: str = Field(...)
    name: str = Field(...)
    status: Optional[str] = Field(None)
    execution_status: Optional[str] = Field(None)
    evaluation_status: Optional[str] = Field(None)
    started_at: Optional[str] = Field(None)
    completed_at: Optional[str] = Field(None)
    duration: Optional[float] = Field(None)
    error_message: Optional[str] = Field(None)
    group_name: Optional[str] = Field(None)
    tags: List[str] = Field(default_factory=list)


class TaskDetailData(APIModel):
    id: int = Field(...)
    name: str = Field(...)
    description: Optional[str] = Field(None)
    status: str = Field(...)
    type: str = Field(...)
    config: Dict[str, Any] = Field(default_factory=dict)
    algorithm_type: Optional[str] = Field(None)
    algorithm_params: Optional[Dict[str, Any]] = Field(None)
    started_at: Optional[str] = Field(None)
    completed_at: Optional[str] = Field(None)
    expected_total_time: Optional[str] = Field(None)
    expected_complete_time: Optional[str] = Field(None)
    used_time: Optional[str] = Field(None)
    total_cases: Optional[int] = Field(None)
    case_count: Optional[int] = Field(None)
    device_count: Optional[int] = Field(None)
    completed_cases: Optional[int] = Field(None)
    failed_cases: Optional[int] = Field(None)
    tags: List[str] = Field(default_factory=list)
    cases: List[TaskCaseBrief] = Field(default_factory=list)
    devices: List[TaskDeviceBrief] = Field(default_factory=list)
    apis: List[TaskApiBrief] = Field(default_factory=list)
    created_at: Optional[str] = Field(None)
    updated_at: Optional[str] = Field(None)


class TaskProgressCurrentCase(APIModel):
    case_id: str = Field(...)
    name: str = Field(...)
    step: Optional[str] = Field(None)
    started_at: Optional[str] = Field(None)


class TaskProgressData(APIModel):
    task_id: str = Field(...)
    status: str = Field(...)
    total_cases: int = Field(...)
    completed_cases: int = Field(...)
    failed_cases: int = Field(...)
    progress: float = Field(...)
    current_case: Optional[TaskProgressCurrentCase] = Field(None)
    updated_at: Optional[str] = Field(None)


class TaskStartData(APIModel):
    task_id: str = Field(...)
    start_time: Any = Field(None)
    status: str = Field(...)
    expected_total_time: Optional[int] = Field(None)
    expected_complete_time: Optional[str] = Field(None)


class TaskUpdateCasesData(APIModel):
    task_id: str = Field(...)
    total_count: int = Field(...)


class TaskStatsData(APIModel):
    total: int = Field(...)
    completed: int = Field(...)
    failed: int = Field(...)
    pending: int = Field(...)
    skipped: int = Field(...)
    pass_rate: float = Field(...)
    tag_stats: Dict[str, Any] = Field(default_factory=dict)
    duration: Any = Field(None)


class TaskCreateParameters(APIModel):
    case_ids: List[str] = Field(default_factory=list)
    device_ids: List[int] = Field(default_factory=list)
    api_ids: List[int] = Field(default_factory=list)
    tags: List[str] = Field(default_factory=list)


class TaskCreateRequest(APIModel):
    name: str = Field(...)
    type: str = Field(...)
    description: Optional[str] = Field(None)
    config: Optional[Dict[str, Any]] = Field(None)
    created_by: Optional[str] = Field(None)
    case_ids: Optional[List[str]] = Field(None)
    device_ids: Optional[List[int]] = Field(None)
    api_ids: Optional[List[int]] = Field(None)
    tags: Optional[List[str]] = Field(None)
    algorithm_type: Optional[str] = Field(None)
    algorithm_params: Optional[Dict[str, Any]] = Field(None)


class TaskControlRequest(APIModel):
    action: str = Field(...)
    case_id: Optional[int] = Field(None)


class TaskUpdateCasesRequest(APIModel):
    action: str = Field(...)
    case_ids: List[int] = Field(default_factory=list)


class TaskBatchActionRequest(APIModel):
    action: str = Field(...)
    task_ids: List[int] = Field(default_factory=list)


class TaskMergeRequest(APIModel):
    task_ids: List[int] = Field(default_factory=list)


class TaskListQuery(APIModel):
    page: int = Field(1)
    per_page: int = Field(10)
    status: Optional[str] = Field(None)
    type: Optional[str] = Field(None)
    algorithm_type: Optional[str] = Field(None)
    search: Optional[str] = Field(None)
    start_date: Optional[str] = Field(None)
    end_date: Optional[str] = Field(None)


class TaskBatchExportQuery(APIModel):
    format: str = Field('json')

File: Intelligent-Audio-TEST/frontend/src/domain/model/task.ts
/**
 * Task 领域模型 —— camelCase
 * 对应后端 api_gateway/schemas/task.py::TaskListItem / TaskDetailData
 */
import type { TaskStatusType, TaskType } from '../enums'

/** 任务关联设备简报 */
export interface TaskDeviceBrief {
  id: number
  name: string
  status?: string
  model?: string
}

/** 任务关联 API 简报 */
export interface TaskApiBrief {
  id: number
  name: string
  status?: string
}

/** 任务关联测试用例简报（TaskCaseBrief → camelCase） */
export interface TaskCaseBrief {
  caseId: string
  name: string
  status?: string
  executionStatus?: string
  evaluationStatus?: string
  startedAt?: string
  completedAt?: string
  duration?: number
  errorMessage?: string
  groupName?: string
  tags?: string[]
}

export interface Task {
  id: number | string
  name: string
  description?: string
  type: TaskType
  status: TaskStatusType
  progress?: number
  config?: Record<string, any>
  algorithmType?: string
  algorithmParams?: Record<string, any>
  startedAt?: string
  completedAt?: string
  /** 预计总耗时（格式化文本，如 "12分钟"） */
  expectedTotalTime?: string | number
  /** 预计完成时间 */
  expectedCompleteTime?: string
  /** 已耗时（格式化文本） */
  usedTime?: string
  totalCases?: number
  caseCount?: number
  deviceCount?: number
  completedCases?: number
  failedCases?: number
  tags?: string[]
  createdAt?: string
  updatedAt?: string
  finishedAt?: string
  error?: string
  deleted?: boolean
  result?: any
  reports?: {
    count: number
    reports: Array<{ id: number; name: string; status: string; type: string; createdAt?: string }>
  }
  devices?: TaskDeviceBrief[]
  apis?: TaskApiBrief[]
  /** 任务详情接口返回的关联测试用例列表（TaskDetailData.cases） */
  cases?: TaskCaseBrief[]
}

/** 任务创建表单/请求（前端 Domain → adapter 转 DTO；对齐 TaskCreateRequest 契约） */
export interface TaskCreateDraft {
  name: string
  type: TaskType
  description?: string
  /** 对应 case_ids: List[str]（后端用例 ID 为字符串） */
  caseIds: string[]
  /** 对应 device_ids: List[int] */
  deviceIds?: number[]
  /** 对应 api_ids: List[int] */
  apiIds?: number[]
  tags?: string[]
  algorithmType?: string
  algorithmParams?: Record<string, any>
  config?: Record<string, any>
}
